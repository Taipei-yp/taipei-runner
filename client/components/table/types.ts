export type TableSort = {
  field: string | null;
  direction: SortDirection | null;
};

export type SortDirection = "asc" | "desc";

export type SortRequestHandler = (field: string) => void;
