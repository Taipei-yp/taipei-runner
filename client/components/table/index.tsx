export { Table } from "./table";
export { SortDirection, TableSort, useNullableTableSort } from "./table-sort";
