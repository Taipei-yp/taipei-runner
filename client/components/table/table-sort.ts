import { useCallback, useState } from "react";

export type TableSort = {
  sortFieldId: string | null;
  direction: SortDirection | null;
};

export type SortDirection = "asc" | "desc";

type SortRequestHandler = (field: string) => void;

// Вообще это можно описать через switch и стейт машину, но кажется тут совсем изевый вариант для этого.
const directionState: Record<SortDirection, SortDirection | null> = {
  asc: "desc",
  desc: null,
};

export function useNullableTableSort(): [TableSort, SortRequestHandler] {
  const [{ direction, sortFieldId }, setSort] = useState({
    // Хорошо бы давать более осмысленные названия переменным
    // Инача от обилия field сложно разобраться в том, что проиходит
    sortFieldId: null,
    direction: null,
  } as TableSort);

  const handleSortRequest = useCallback(
    (fieldId: string) => {
      // updating sort: null -> asc -> desc -> null -> asc -> ...
      const newSortFieldId = direction === "desc" ? null : fieldId;
      const newDirection = direction ? directionState[direction] : "asc";
      setSort({ sortFieldId: newSortFieldId, direction: newDirection });
    },
    [direction],
  );

  return [{ direction, sortFieldId }, handleSortRequest];
}

// we can add other variants here, maybe
