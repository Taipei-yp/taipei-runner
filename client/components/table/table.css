.table {
  width: 100%;
  border-collapse: collapse;

  td {
    padding: 15px;
  }

  thead {
    color: var(--tpr-color-accent);
    font-size: var(--tpr-fontsize-xl);

    td {
      cursor: pointer;
    }
  }

  tbody {
    color: var(--tpr-color-primary);
    font-size: var(--tpr-fontsize-l);
  }

  tr {
    height: calc(2 * var(--tpr-fontsize-xl));
  }

  &__sort-direction {
    font-size: 0.83em;
  }
}
