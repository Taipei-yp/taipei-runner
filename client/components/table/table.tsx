import block from "bem-cn";
import React, { ComponentType, memo, MouseEvent, useCallback } from "react";
import { TableSort } from "client/utils/ui-types";

import "./table.css";

const b = block("table");

type Props<TData extends WithId> = {
  className?: string;
  headers: Array<HeaderInfo<TData>>;
  data: TData[];
  components?: Partial<Record<keyof TData, InnerComponent>>;
  sort?: TableSort;
  onHeaderClick?: (field: string) => void;
};

type HeaderInfo<TData extends WithId> = {
  title: string;
  field: keyof TData;
};

type WithId = {
  id: unknown;
  [key: string]: unknown;
};

// TODO: research react component generic types
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type InnerComponent = ComponentType<{ value: any }>;

const TextRenderer = (({ value }) => value) as InnerComponent;

const Table = <TData extends WithId>({
  className,
  headers,
  data,
  components = {},
  sort = { field: null, direction: null },
  onHeaderClick = () => {},
}: Props<TData>) => {
  const allComponents = headers.reduce((acc, { field }) => {
    acc[field] = components[field] ?? TextRenderer;
    return acc;
  }, {} as Record<keyof TData, InnerComponent>);

  let DataComponent: InnerComponent;

  const handleHeadClick = useCallback(
    (event: MouseEvent) => {
      const { field } = (event.target as HTMLTableSectionElement).dataset;
      onHeaderClick(field!);
    },
    [onHeaderClick],
  );

  return (
    <table className={b.mix(className)}>
      <thead>
        <tr onClick={handleHeadClick}>
          {headers.map(header => (
            <td key={header.field as string} data-field={header.field}>
              {header.title}
              {sort.field === header.field && (
                <span className={b("sort-direction")}>
                  {sort.direction === "asc" ? "▼" : "▲"}
                </span>
              )}
            </td>
          ))}
        </tr>
      </thead>
      <tbody>
        {data.map(datum => (
          <tr key={datum.id as string}>
            {headers.map(header => (
              <td key={`${datum.id}-${header.field}`}>
                {
                  ((DataComponent = allComponents[header.field]),
                  (<DataComponent value={datum[header.field]} />))
                }
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
};

const WrappedTable = memo(Table);
export { WrappedTable as Table };
