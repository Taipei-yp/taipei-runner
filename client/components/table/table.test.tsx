import React from "react";
import TestRenderer from "react-test-renderer";
import { Table } from "client/components/table/table";

describe("Component Table", () => {
  it("renders table by data and headers", () => {
    const headers = [
      {
        title: "FIRST-FIELD",
        field: "firstField",
      },
      {
        title: "SECOND-FIELD",
        field: "secondField",
      },
    ];
    const data = [
      {
        id: 0,
        firstField: 1,
        secondField: "1",
      },
      {
        id: 1,
        firstField: 2,
        secondField: "2",
      },
      {
        id: 2,
        firstField: 3,
        secondField: "3",
      },
    ];
    const component = TestRenderer.create(
      <Table headers={headers} data={data} />,
    );
    expect(component).toMatchSnapshot();
  });

  it("renders empty table for empty data", () => {
    const component = TestRenderer.create(<Table headers={[]} data={[]} />);
    expect(component).toMatchSnapshot();
  });
});
