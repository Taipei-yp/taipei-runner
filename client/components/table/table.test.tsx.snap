// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Table renders empty table for empty data 1`] = `
<table
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <thead>
    <tr
      onClick={[Function]}
    />
  </thead>
  <tbody />
</table>
`;

exports[`Component Table renders table by data and headers 1`] = `
<table
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <thead>
    <tr
      onClick={[Function]}
    >
      <td
        data-field="firstField"
      >
        FIRST-FIELD
      </td>
      <td
        data-field="secondField"
      >
        SECOND-FIELD
      </td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        1
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        2
      </td>
    </tr>
    <tr>
      <td>
        3
      </td>
      <td>
        3
      </td>
    </tr>
  </tbody>
</table>
`;
